Automattic/jetpack
# Jetpack AI: redirect user back to editor after buying upgrade

Relates to #34946.

## Proposed changes:

* Add support for a `redirect_to_referrer` parameter on the My Jetpack product interstitial page; when present, the code will look for the referrer information and use it as the redirect URL after the checkout
* Include `redirect_to_referrer` parameter on the Jetpack AI upgrade buttons for Jetpack sites

The final result is: 

- on Jetpack sites, when a Jetpack AI upgrade button is clicked from inside the editor, the user will be redirected back to the editor after the purchase. Same for the site editor

Note: this PR does not address the issue on Simple/Atomic sites for now. I will be tackling it on a follow up.

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






* Test on a Jetpack site that does not have the legacy Jetpack AI plan (because you will not see the upgrade button on them)
* Create a new post and add some content
* On the editor, look for the Jetpack sidebar, the "AI Assistant" section, and click the "upgrade" button to start the upgrade flow
   * Notice the URL on the editor right before the redirect happens; this is the URL you will land after upgrading the plan
* Confirm you landed on the Jetpack AI interstitial and that the URL contains a `redirect_to_referrer` parameter:

<img width="600" alt="Screenshot 2024-01-29 at 15 45 29" src="https://github.com/Automattic/jetpack/assets/6760046/a67adb44-c530-49cb-8577-f946e6282166">

* Click the upgrade button on the Jetpack AI interstitial
* Confirm you landed on the checkout page and that the `redirect_to` parameter contains the URL you were before starting the upgrade flow:

<img width="500" alt="Screenshot 2024-01-29 at 15 47 07" src="https://github.com/Automattic/jetpack/assets/6760046/6a7449d4-5a27-42b3-8ab5-0f57cb6ed702">

* Proceed with checkout
* After buying the new tier, confirm you are redirected back to the editor, to the same post you were editing before

---

Other test scenarios for Jetpack:

- start a new post but do not add any content to it, then start the upgrade flow; you should land back to an empty editor
- edit a post you already have
- execute requests up to the limit of the current tier, then try to use the AI Assistant on the site editor; when the upgrade prompt appears, follow the checkout path and make sure you land back to the site editor after the purchase
